src/coreclr/vm/prestub.cpp

```diff
@@ -2053,6 +2050,23 @@ SigPointer MethodDesc::GetAsync2ThunkResultTypeSig()
     // ReturnType comes now. Skip the modifiers (like async2 modifier).
     IfFailThrow(pSig.SkipCustomModifiers());
 
+    CorElementType etype;
+    IfFailThrow(pSig.PeekElemType(&etype));
+
+    if ((etype & ELEMENT_TYPE_GENERICINST) != 0)
```

Comment:
```suggestion
    if (etype == ELEMENT_TYPE_GENERICINST)
```
This is not a flag.

Reply:
Also, can it be an assert? I think it should be impossible for this not to be a `Task<T>` or `ValueTask<T>` instantiation here.

Reply:
Initially I thought of trying to keep the capability of parsing the old (modreq'd) signatures - in case we need that for some internal reasons, but checking for `ELEMENT_TYPE_GENERICINST` here is not the way. I think it would not work if we have a modreq'd signature that returns `Task<T>` as we would unwrap both modreq and Task wrappers. 

Yes. It would be simpler if this is just an assert. 
If we need to parse something else (I have more and more doubts) we would just have a different helper.